compiler/module_graph.rs

```diff
@@ -0,0 +1,1022 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+use crate::ast::parse;
+use crate::ast::Location;
+use crate::ast::ParsedModule;
+use crate::compiler::CompileOptions;
+use crate::compiler::CompilerEmit;
+use crate::compiler::CompilerIsolate;
+use crate::compiler::TranspileOptions;
+use crate::import_map::ImportMap;
+use crate::msg::EmittedFile;
```

Comment:
Note to self: `ModuleGraph`/`Graph` should be used for `deno info`, investigate if this struct could be made into two separate structs - one for loading and parsing of dependencies (think collecting module tree), second for integration with TS compiler cache (think utilizing the graph that collects module tree to process the tree further). See https://github.com/denoland/deno/pull/6786 for reference

Reply:
While it sort of overloads it a bit, I sort of like them together... In that `Graph` contains all the information that ops need to access files, you just wire it into that...  I like having all the "what you can do with a graph" on the surface API of the graph...

For `GraphBuilder` there is only:

- insert - Add a specifier as a "root".
- get_graph - Move the graph out.

For `Graph` it currently is:

Updating a graph:
- compile (update the graph with type checked code)
- transpile (update the graph with type stripped code)

Things I need to refactor out/make non-public:
- update (take emitted files and update the graph... I need to refactor this out, as compile and transpile should automatically do this, making the compiler less dependent)
- flush (flush any "dirty" modules to the cache, again this should be refactored out as it need to be handled by compiler and transpile)
- get_root_names (this was used by compiler.compile but needs to be passed)
- get_sources (this was used by transpile but need to be passed)

Then needed for compiler ops:

- pub modules (the Hashmap of modules)
- resolve (takes a string specifier and a base module specifier and returns the module specifier in the graph, abstracting out all the logic for `@deno-types` and `X-TypeScript-Types` and triple-slash reference)
- get_build_info/set_build_info

But talking that through here sounds like I should refactor that last group as a trait (and abstract out the public access to the hashmap of module).

Adding in `info()` to the `Graph` makes a lot of sense, and/or have the Display of a Graph be the output we need for info.

Reply:
Ok, I have done a refactor... I still want to do some fine tuning of it, but basically the compiler and the graph are better decoupled, and the compiler ops only depend upon to methods in a trait.  It means in practice, the only thing that needs to be interfaced with is the `GraphBuilder` and the `Graph`.